src/sonic-ctrmgrd/ctrmgr/container_state.py

```diff
@@ -0,0 +1,291 @@
+    ct = ct_version.split('.') if ct_version else "0.0.0".split('.')
+    nxt = version.split('.') if version else "0.0.0".split('.')
+    ret = False
+    for cs, ns in zip(ct, nxt):
+        c = int(cs)
+        n = int(ns)
+        if n < c:
+            ret = True
+            break
+        elif n > c:
+            break
+
+    debug_msg("compare version: new:{} current:{} res={}".format(
+        version, ct_version, ret))
+    return ret
+
+
+def is_active(feature):
+    if state_data[SYSTEM_STATE] == "up":
+        return True
+    else:
+        syslog.syslog(syslog.LOG_ERR, "Found inactive for {}".format(feature))
+        return False
+
+
+def update_state(is_up, feature, owner=None, version=None):
+    data = {
+            CURRENT_OWNER: owner if is_up else "none",
+            DOCKER_ID: get_docker_id() if is_up else "",
+            UPD_TIMESTAMP: str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+            }
+    if is_up:
+        data[VERSION] = version
+
+        if (owner == "local"):
+            # Disable deployment of this version as available locally
+            drop_label(feature, version)
+        else:
+            data[REMOTE_STATE] = "running"
+
+    elif state_data[CURRENT_OWNER] != "local":
+        state = state_data[REMOTE_STATE]
+        if state != "pending":
+            if state != "running":
+                syslog.syslog(syslog.LOG_ERR,
+                        "{} kube down state: {} != running", feature, state)
+            else:
+                data[REMOTE_STATE] = "stopped"
+
+    debug_msg("{} up= {} data:{}".format(feature, is_up, str(data)))
+    update_data(feature,  data)
+    state_data.update(data)
+
+
+def do_exit(feat, m):
+    # Exiting will kick off the container to run.
+    # So sleep forever with periodic logs.
+    #
+    while True:
+        syslog.syslog(syslog.LOG_ERR, "Exiting .... feat:{} docker_id:{} msg:{}".format(
+            feat, get_docker_id(), m))
+        time.sleep(60)
+    
+
+def container_up(feature, owner, version):
+    debug_msg("BEGIN")
+    read_data(feature)
+
+    debug_msg("args: feature={}, owner={}, version={} DB: set_owner={} state_data={}".format(
+        feature, owner, version, set_owner, state_data))
+
+    if owner == "local":
+        update_state(True, feature, owner, version)
+    else:
+        if (set_owner == "local"):
+            do_exit(feature, "bail out as set_owner is local")
+
+        if not is_active(feature):
+            do_exit(feature, "bail out as system state not active")
+
+        if check_unset_labels(feature, version):
+            do_exit(feature, "This version is marked disabled. Exiting ...")
+
+        if instance_lower(feature, version):
+            # Remove label <feature_name>_<version>_enabled
+            # Else kubelet will continue to re-deploy every 5 mins, until
+            # master removes the lable to un-deploy.
+            #
+            do_exit(feature, "bail out as current deploy id is lower")
```

Comment:
where are these logic described in the design doc? i am not able to find it. Here is what I see.  is this the right place? but it seems you are doing much more than the actions described below.

```
On post-start

current_owner = local/kube
docker_id = <ID of the container>
The start.sh of the container (called from supervisord) is updated to call system container state <name> up <kube/local>, which in turn would do the above update. The application process is started after the call to container_state script.
```

Reply:
The design doc was explaining the normal run, but  The above are for catching the *possible* failure scenarios.

In all cases above, except "instance_lower", the kube would not have deployed, if API server was reachable. Because, when we switch owner to local / systemctl stop / explicitly-disable-a-version, we set/drop appropriate labels. If API server was reachable, then it would have instructed kubelet in switch *not-to-deploy*. But if API server is not reachable, the kubelet would attempt to re-deploy (as per old instructions) and it gets blocked here.

The case of instance_lower could be a mistake or by design (like we showed in demo & PPT), where the lower one is blocked.

We could potentially add a section to HLD (_guess it would be a good value too_) to explain switch behavior in different scenarios, when API server become unreachable.

Reply:
Updated doc -- [here](https://github.com/Azure/SONiC/blob/b0b6080712d0a27c0a4748e921972901a8e3ab65/doc/kubernetes/Kubernetes-support.md#kubernetes-cluster-unreachable-impacts)